src/context.rs

```diff
@@ -1,88 +1,360 @@
+/// }
+///
+/// impl<T, C> hyper::server::Service for MiddlewareService<T, C>
+///     where
+///         C: Pop<MyItem>,
+///         T: hyper::server::Service<Request = (hyper::Request, C::Result)>
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = T::Response;
+///     type Error = T::Error;
+///     type Future = T::Future;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         let (item, context) = context.pop();
+///         do_something_with_my_item(&item);
+///         self.inner.call((req, context))
+///     }
+/// }
+///
+/// # fn main() {}
+pub trait Pop<T> {
+    /// The type that remains after the value has been popped.
+    type Result;
+    /// Extracts a value.
+    fn pop(self) -> (T, Self::Result);
+}
 
-    /// Set the context logger
-    fn set_logger(&mut self, logger: slog::Logger);
+/// Defines a method for inserting a value, changing the resulting
+/// type. Used to specify that a hyper service adds some data from the context,
+/// making it available to later layers, e.g.
+///
+/// ```rust
+/// # extern crate hyper;
+/// # extern crate swagger;
+/// # extern crate futures;
+/// #
+/// # use swagger::context::*;
+/// # use futures::future::{Future, ok};
+/// # use std::marker::PhantomData;
+/// #
+/// # struct MyItem;
+/// #
+/// struct MiddlewareService<T, C> {
+///     inner: T,
+///     marker: PhantomData<C>,
+/// }
+///
+/// impl<T, C> hyper::server::Service for MiddlewareService<T, C>
+///     where
+///         C: Push<MyItem>,
+///         T: hyper::server::Service<Request = (hyper::Request, C::Result)>
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = T::Response;
+///     type Error = T::Error;
+///     type Future = T::Future;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         let context = context.push(MyItem{});
+///         self.inner.call((req, context))
+///     }
+/// }
+///
+/// # fn main() {}
+pub trait Push<T> {
+    /// The type that results from adding an item.
+    type Result;
+    /// Inserts a value.
+    fn push(self, T) -> Self::Result;
 }
 
-impl HasLogger for Context {
-    fn logger(&self) -> &Option<slog::Logger> {
-        &self.logger
-    }
+/// Defines a struct that can be used to build up contexts recursively by
+/// adding one item to the context at a time, and a unit struct representing an
+/// empty context. The first argument is the name of the newly defined context struct
+/// that is used to add an item to the context, the second argument is the name of
+/// the empty context struct, and subsequent arguments are the types
+/// that can be stored in contexts built using these struct.
+///
+/// A cons list built using the generated context type will implement Has<T>
+/// for each type T that appears in the list, provided that the list only
+/// contains the types that were passed to the macro invocation after the context
+/// type name.
```

Comment:
Sorry, that documentation is a bit unclear, and it doesn't help that some of it hasn't been updated.

The macro invocation
```
new_context_type!(MyContext, MyEmpContext, MyType1, MyType2, MyType3);
```
will define new types `MyContext<T, C>` and `MyEmpContext` that you can use to build list types like so:
```
MyContext<MyType1, MyContext<MyType2, MyEmpContext>>
```
This is the type of a list of length 2 containing a `MyType1` and a `MyType2`. It will implement:

- `Has<MyType1> and `Has<MyType2>`, because these types both appear in the list itself, *and* were passed to the macro invocation
- `Pop<MyType1> and `Pop<MyType2>` for the same reason

It won't implement `Has<MyType3>` or `Pop<MyType3>` because it doesn't contain a value of that type.

If we did
```
MyContext<MyType1, MyContext<MyType4, MyEmpcontext>>
```
this *won't* implement `Has<MyType4>` even though it appears in the list, because that type wasn't passed to the macro invocation that defined `MyContext`.

All of these will implement `Push<T>` for any `T`. But it's only a good idea to use it for the types passed to macro invocation, since otherwise it might not be possible to get the values out again.

Is that a bit clearer?